### ISSUE TITLE
Training and testing with custom augmented data 

### ISSUE THREAD

**Issue by abhigoku10** [dbolya/yolact#81]
@dbolya Hi , i trained the darknet model with augmentation which i had performed externally without using the internal augmentation . The accuracy of the model was acceptable and the loss value was around 8-9 value but the performance of the model dropped to half compared to model without augmentation , theoretically by adding extra augmented data should improve the accuracy but the fps of the network should remain the same right , any thoughts on this

**Comment by dbolya** [comment]
We already augment the data _a lot_, so I don't think there's going to be any benefit from external augmentation. And it isn't always true that the more augmentation the better, since it's easy to destroy your data if you pile on too much data augmentation. You want to have your network trained on as closes to the real data as possible, so all your augmentations should create new images that look as if they were from the original dataset.

**Comment by abhigoku10** [comment]
@dbolya  yes rightly said , i am bit less concerned with the accuracy my problem is the fps is dropping when augmented data is added which is strange from my end

**Comment by dbolya** [comment]
You mean the eval FPS after training goes down when you add more data? That is a little bit odd, but it is possible, since the FPS is tied to performance:

Say you had a model that performed perfectly. Then, it would output only one high-scoring detection for each object in the image. That means there will only be a few detections going into NMS and the same few detections having their masks generated. 

Now imagine a really poor performing model. Such a model might have way too many high scoring detections, meaning a lot of detections will reach NMS. Then say the model is really bad so none of these detections are actually overlapping, so most of them survive NMS. Then all 100 of them (per COCO standards we output a max of 100 detections) have their masks generated and worked on by postprocessing. That's potentially 10-20x the work for postprocessing compared to the perfect model.

The actual FPS drop likely depends on how well your hardware can support more detections, but YOLACT is pretty unique in that it doesn't do a lot of per detection post processing, and all that per detection stuff is vectorized (not a single for loop over num_dets during eval!). How large was this FPS drop? Do you think this is enough to explain it?

**Comment by abhigoku10** [comment]
@dbolya  thanks for the elaborate explanation , the model trained with augmented data i tested it on few images and it was working good but when i checked the fps it changed drastically from 18fps to 9 fps on a high end 2080ti gpu . yes i am talking about eval fps

**Comment by dbolya** [comment]
Okay, if the images looked good then the FPS decrease is cause for concern. All you did was add images to the training set by modifying the annotations file, right? You didn't edit the code at all?

**Comment by abhigoku10** [comment]
@dbolya  yes i just added the images no part of the code was edited

**Comment by dbolya** [comment]
Then I'm not sure what could have gone wrong. Are you able to reproduce the 18 fps of the old model with your environment? Maybe something there changed?

**Comment by abhigoku10** [comment]
@dbolya i shall double check this and keep u posted on this activity

**Comment by abhigoku10** [comment]
@dbolya i double checked the issues and i was able to solve it , it was due to oversight modifications on the config file . Thanks for the support

**Comment by dbolya** [comment]
Oh nice! I'll close the issue then.

**Comment by abhigoku10** [comment]
@dbolya  for my custom data should i change the anchor box sizes since my custom data consists of small objects

**Comment by dbolya** [comment]
Yeah, that would be good to do. If you're training on something like cityscapes, I'd half all the anchor box scales.

**Comment by abhigoku10** [comment]
@dbolya steps to generate the anchor boxes for the custom dataset?

**Comment by dbolya** [comment]
I mean just open and read your dataset, and average together all the bbox areas. Take the sqrt to find the averages scale. Then do the same for COCO. Then divide your average scale by COCO's average scale. Finally, multiply all the scales in the backbone config with this ratio.

**Comment by AndreiBaraian** [comment]
What parameter does changes the anchor box sizes? Is it pred_scales?

**Comment by dbolya** [comment]
@AndreiBaraian Yeah pred_scales controls the scale of the boxes (`sqrt(w * h)`), while pred_aspect_ratios controls the aspect ratios (`w / h`). So to get a specific w and h you can do the math based on that.
